Document how to run uvicorn programatically (#1525)
Kludex/uvicorn: docs/index.md

### Running programmatically

> There are many ways to run uvicorn directly from your application.

Reviewer: "Many" feels a bit overwhelming — as we document here, there are only 3 ways (run / config-serve.run() / config-serve.serve()).

```suggestion
There are several ways to run uvicorn directly from your application.
```

> #### `uvicorn.run` function

Reviewer: Aiming at concise headings

```suggestion
#### `uvicorn.run`
```

> The easiest way is to run via `uvicorn.run` function:

Reviewer: "Easiest" is subjective unfortunately; here's a wording trying to more explicitly convey what use cases this usage could address:

```suggestion
If you're looking for a programmatic equivalent of the `uvicorn` command line interface, use `uvicorn.run()`:
```

(It does accept all options that `$ uvicorn [OPTIONS]` does, correct?)

Author: It does. 👍

> ```python

Author: ```suggestion
```python
# main.py
```

import uvicorn

async def app(scope, receive, send):
    ...

if __name__ == "__main__":
    uvicorn.run("main:app", host="127.0.0.1", port=5000, log_level="info")

> ```

Reviewer: ```markdown
`uvicorn.run` accepts the following arguments:

* `app`: Either an ASGI application, or (as shown above) an import string to the target ASGI application. The latter is required if using auto-reload or multiple workers.
* `**kwargs`: Keyword arguments that mirror command line options (see [Settings](/settings.md)): `--host <str>` becomes `host=<str>`, `--log-level <str>` becomes `log_level=<str>`, etc.
```

(I think this kind of info is valuable in docs, but maybe we might want to kick off an `API Reference` page for this kind of thing. I'm OK if we defer this bit from this PR.)

Author: We _improved_ the `uvicorn.run` type annotation, so we can have autocomplete on it: https://github.com/encode/uvicorn/blob/b1a45821a2ab916b4428bdcb9eba4a4a0cc8fb74/uvicorn/main.py#L445-L492

I'll not apply this for now. An API reference would be fine by me.

> #### Explicitly setup `Config` and `Server` instances

Reviewer: ```suggestion
#### `Config` and `Server` instances
```

> You can create the `uvicorn.Config` and `uvicorn.Server` instances manually, and then use `uvicorn.Server.run()`:

Reviewer: ```suggestion
For more control over configuration and server lifecycle, use `uvicorn.Config` and `uvicorn.Server`:
```

    ...

if __name__ == "__main__":
    config = uvicorn.Config("main:app", port=5000, log_level="info")
    server = uvicorn.Server(config)
    server.run()
```

If you'd like to run Uvicorn from an already running async environment, use `uvicorn.Server.serve()` instead:

> ```python

Reviewer 2: ```suggestion
main.py
```python
```

Reviewer 2: just to say this is the name of the file, I'll approve it without this change but we have too many people not understanding how the app string works

Author: Added on the first file here.

    ...

if __name__ == "__main__":
    config = uvicorn.Config("main:app", host="127.0.0.1", post=5000, log_level="info")
    server = uvicorn.Server(config)
    server.run()
```

#### Run inside the event loop

There are cases that `async` logic needs to be performed, and it's interesting to run `uvicorn` inside the event loop:

```python
import asyncio
import uvicorn

async def app(scope, receive, send):
    ...

async def main():
    config = uvicorn.Config("main:app", host="127.0.0.1", post=5000, log_level="info")
    server = uvicorn.Server(config)
    await server.serve()

if __name__ == "__main__":

>     asyncio.run(main())

Reviewer: I'd be tempted to merge this section with "Config and Server instances" since it describes an alternate way of using these -- .run() for launching an async environment, .serve() if you've already got one running.

It might also be that we don't need any h4 here at all (I'm looking at the table of contents sidebar of this home page which doesn't contain any h4 yet).

```suggestion
If you'd like to run Uvicorn from an already running async environment, use `uvicorn.Server.serve()` instead:

```python
import asyncio
import uvicorn

async def app(scope, receive, send):
    ...

async def main():
    config = uvicorn.Config("main:app", host="127.0.0.1", post=5000, log_level="info")
    server = uvicorn.Server(config)
    await server.serve()

if __name__ == "__main__":
    asyncio.run(main())
```
